book/en/src/internals/targets.md

```diff
@@ -0,0 +1,59 @@
+# Target Architecture
+
+While RTIC can currently target all Cortex-m devices there are some key architecure differences that 
+users should be aware. Namely the absence of hardware priority ceiling (BASEPRI) support in the 
+ARMv6-M and ARMv8-M-base architectures requires a few tweaks from RTIC to deliver the same 
+features. These differences result in two flavors of critical sections: priority ceiling, and source 
+masking. Table 1 below shows a list of Cortex-m processors and which type of critical section they 
+employ.
+
+#### *Table 1: Critical Section Implementation by Processor Architecture*
+
+| Processor  | Architecture | Priority Ceiling | Source Masking |
+| :--------- | :----------: | :--------------: | :------------: |
+| Cortex-M0  | ARMv6-M      |                  |     &#2713     |
+| Cortex-M0+ | ARMv6-M      |                  |     &#2713     |
+| Cortex-M3  | ARMv7-M      |      &#2713      |                |
+| Cortex-M4  | ARMv7-M      |      &#2713      |                |
+| Cortex-M23 | ARMv8-M-base |                  |     &#2713     |
+| Cortex-M33 | ARMv8-M-main |      &#2713      |                |
+| Cortex-M7  | ARMv7-M      |      &#2713      |                |
+
+## Priority Ceiling
+
+This implementation is covered in depth by Chapter 4.5 of this book.
+
+## Source Masking
+
+Since there is no hardware support for a priority ceiling, RTIC must instead rely on the Nested 
+Vectored Interrupt Controller (NVIC) present in the core architecture. Consider Figure 1 below, 
+showing two tasks A and B where A has higher priority but shares a resource with B. 
+
+#### *Figure 1: Shared Resources and Source Masking*
+
+```text
+  ┌────────────────────────────────────────────────────────────────┐
+  │                                                                │
+  │                                                                │
+3 │                   Pending    Preempts                          │
+2 │             ↑- - -A- - - - -↓A─────────►                       │
+1 │          B───────────────────► - - - - B────────►              │
+0 │Idle┌─────►                             Resumes  ┌────────►     │
+  ├────┴────────────────────────────────────────────┴──────────────┤
+  │                                                                │
+  └────────────────────────────────────────────────────────────────┴──► Time
+                t1    t2        t3         t4
+```
+
+At time *t1*, task B locks the shared resource by selectively disabling all other tasks which share 
+the resource using the NVIC. In effect this raisis the virtual priority ceiling. Task A is one such
+task that shares resources with task B. At time *t2*, task A is either spawned by task B or becomes 
+pending through an interrupt condition, but does not yet preempt task B even though it's priority is 
+greater. This is because the NVIC is preventing it from starting due to task A's source mask being 
```

Comment:
```suggestion
greater. This is because the NVIC is preventing it from starting due to task A being _masked_.
```

Reply:
Great to see that the "under the hood" section gets some love! Thanks all for your contributions.

Some general comments:
- RTIC implements SRP (or rather SRP with fixed/static priorities).
- For the cortex-m architecture we have two implementations thereof
  - BASEPRI (for those architectures that provides BASEPRI), and for other cases
  - source masking.
   
Both approaches achieves the same thing, i.e., raising the current SRP system ceiling once a resource is locked. Notice here, source masking disables ALL tasks with priority equal or less than the current SRP system ceiling (not just the tasks using the particular resource being locked). This way RTIC ensures the singe blocking property of SRP. Although a small detail this has huge implications to the response time guarantees as transitive blocking is a can of worms when it comes to guarantees.

As an example, if just disabling task A in the example, any task N (with a priority larger than B less/equal A) could preempt B (while B holds the shared resource) and thus further block A from starting. 

This can NEVER happen under SRP, and as RTIC implements SRP, it will NEVER happen using RTIC.

The proposed text is mostly correct. It should be made clear that both approaches implements SRP, where the current system ceiling is implemented either using BASEPRI or source masking.

A technical detail is that (at least for the M0/M0+) the maximum number of interrupt sources/vectors is 32, thus a single (32 bit)  mask register is sufficient. This makes the source masking atomic and almost as efficient as the BASEPRI implementation. The difference is that setting a BASEPRI regisiter is done with a core register write (single instruction), while manipulating the NVIC requires the NVIC_ISER/NVIC_ICER address to be held in a register (at the cost of setting up the address and slight increase of "register pressure" for the LLVM backend).

Reply:
I'm not sure if this is related to "source mask" vs "interrupt enable" semantics, so I quoted it below so we don't forget about it once this item is closed.